crypto/src/symkey/chacha20poly1305/mod.rs

```diff
@@ -0,0 +1,308 @@
+use crate::random::{CryptoRng, Rng};
+use crate::symkey::Error;
+use chacha20poly1305::aead::{AeadInPlace, NewAead};
+use chacha20poly1305::{XChaCha20Poly1305, XNonce};
+use parity_scale_codec::{Decode, Encode};
+
+pub const NONCE_LEN: usize = 24;
+pub const KEY_LEN: usize = 32;
+
+#[must_use]
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Chacha20poly1305Key {
+    key_data: chacha20poly1305::Key,
```

Comment:
One thing here is the key isn't zeroed on drop. As in, in the lib they don't do it

Reply:
https://docs.rs/aead/0.4.3/src/aead/lib.rs.html#79

Reply:
I have some questions. What do you guys think about having `Clone + Debug` implemented for the private key? Are there any security concerns? I don't know if `Clone` can be harmful in any way but maybe malicious actor can somehow trigger the key to be printed to log at which point zeroing the memory is not really as effective.

This maybe useful for zeroing memory on drop: https://docs.rs/zeroize/latest/zeroize/#custom-derive-support

Reply:
I know these limitations, but there's almost nothing we can do currently as this is a whole campaign on its own. The right thing to do is to add mlock support, but it's POSIX specific and I found one esoteric crate that uses it and I'm not sure how it works on Windows. Windows has its its own VirtualLock... so imagine the nightmare of implementing platform specific things. 

On one hand, Copy can be an issue, but clone is necessary for the struct to be useful; after all it's just an array, and the wallet may require to clone the key. The problem with copy, AFAIU, is that it triggers intermediate objects all over the place. Clone is strict and will only create a copy when asked to. The clones will have the same characteristics. 

Now about Debug, maybe we can remove it. But we have to keep something in mind: If a computer is compromised by a "malicious actor" to the point of controlling the program, then everything is already compromised. The actor can already take a copy of the virtual memory of the program and look into it. In other words, we want to be careful of overdoing things.

Reply:
@muursh I think you quoted an older version of the crate. The current version does zeroize:

https://github.com/RustCrypto/AEADs/blob/d93b4e098cd7dfaa72518286060ad0d6ab94f346/chacha20poly1305/src/lib.rs#L311

Reply:
Unresolve the comment if I'm wrong

Reply:
So it does. Not sure how I managed that